### PR TITLE
BYON front end code.

### PR DESCRIPTION
**Pull request by dlabaj**
This PR adds the following functionality to the UI:

- [x] Import BYON Image
- [x] Adds BYON Table
- [x] Adds Notebook Images setting
- [x] Adds ability to delete notebook images
- [x] Adds ability to edit notebook images
- [x] Allows user to turn on and off the visibility of a notebook image.
- [x] Integrates with BYON Rest API.
Screen Shot 2022-04-07 at 10 29 52 AM

Screen Shot 2022-04-07 at 10 30 48 AM
Screen Shot 2022-04-07 at 10 31 04 AM
Screen Shot 2022-04-07 at 10 30 37 AM
Screen Shot 2022-04-05 at 1 56 35 PM

### CODE REVIEW

**Comment by dlabaj** [comment]
CC/ @tumido

File: .env [hunk 1]
```diff
@@ -15,3 +15,4 @@ SOURCE_REPOSITORY_REF=master
 DOC_LINK ='https://opendatahub.io/docs.html'
 COMMUNITY_LINK ='https://opendatahub.io/community.html'
 ENABLED_APPS_CM = 'odh-enabled-applications-config'
+ALLOW_ADMIN=true
```
Comment (tumido) [inline .env:18]: Is this used anywhere? I see no reference to it in the PR.

File: frontend/src/utilities/time.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  } else if (elapsed < msPerHour) {
+    return Math.round(elapsed / msPerMinute) + ' minutes ago';
+  } else if (elapsed < msPerDay) {
+    return Math.round(elapsed / msPerHour) + ' hours ago';
+  } else if (elapsed < msPerMonth) {
+    return 'approximately ' + Math.round(elapsed / msPerDay) + ' days ago';
+  } else if (elapsed < msPerYear) {
+    return 'approximately ' + Math.round(elapsed / msPerMonth) + ' months ago';
+  } else {
+    return 'approximately ' + Math.round(elapsed / msPerYear) + ' years ago';
```
Comment (tumido) [inline frontend/src/utilities/time.ts:21]: When the image is just imported (in `Importing` phase), the reported `notebokk.uploaded` is `undefined`. See here: https://github.com/opendatahub-io/odh-dashboard/pull/171/files#diff-5128194b7438d4f34d77a69aa1b358cb72d24ed2762f816243148ccf83bb3edeR21

This else branch will be probably hit. For example Openshift console defaults to `Now` in such cases.

File: frontend/src/utilities/time.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  const elapsed = current - previous;
+
+  if (elapsed < msPerMinute) {
+    return Math.round(elapsed / 1000) + ' seconds ago';
+  } else if (elapsed < msPerHour) {
+    return Math.round(elapsed / msPerMinute) + ' minutes ago';
+  } else if (elapsed < msPerDay) {
+    return Math.round(elapsed / msPerHour) + ' hours ago';
+  } else if (elapsed < msPerMonth) {
+    return 'approximately ' + Math.round(elapsed / msPerDay) + ' days ago';
```
Comment (tumido) [inline frontend/src/utilities/time.ts:17]: Not sure if the `approximately` word is not redundant/too much for UX. I think stating `7 days ago` is already understood as an approximation. It's shorter and will fit on a single line easier than `approximately 7 days ago` as shown on the screenshots.

File: frontend/src/utilities/useWatchNotebookImages.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  //           setLoadError(undefined);
+  //           setComponents(updatedComponents);
+  //         })
+  //         .catch((e) => {
+  //           setLoadError(e);
+  //         });
+  //     }
+  //   }, [forceUpdate, installed]);
+
+  //   const retComponents = useDeepCompareMemoize<OdhApplication[]>(components);
```
Comment (tumido) [inline frontend/src/utilities/useWatchNotebookImages.tsx:71]: dead code

File: frontend/src/types.ts [hunk 5]
```diff
@@ -116,15 +116,29 @@ export type NotebookError = {
...
+};
+
+export type NotebookUpdateRequest = {
+  id: string;
+  name?: string;
+  description?: string;
   visible?: boolean;
   packages?: NotebookPackage[];
-  error?: NotebookError;
+  software?: NotebookPackage[];
```
Comment (tumido) [inline frontend/src/types.ts:141]: This reverts my latest PR. Update request should not allow `software`.

File: frontend/src/services/notebookImageService.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+//   const options = { params: updateParams };
+//   return axios
+//     .get(url, options)
+//     .then((response) => {
+//       return response.data;
+//     })
+//     .catch((e) => {
+//       throw new Error(e.response.data.message);
+//     });
+// };
```
Comment (tumido) [inline frontend/src/services/notebookImageService.ts:69]: dead code

**Review by LaVLaS — APPROVED** [review]
/lgtm

File: frontend/src/services/notebookImageService.ts (hunk 6 above)
Comment (dlabaj) [reply frontend/src/services/notebookImageService.ts:69]: Done.

File: frontend/src/types.ts (hunk 5 above)
Comment (dlabaj) [reply frontend/src/types.ts:141]: Done

File: frontend/src/utilities/useWatchNotebookImages.tsx (hunk 4 above)
Comment (dlabaj) [reply frontend/src/utilities/useWatchNotebookImages.tsx:71]: Done.

File: frontend/src/utilities/time.ts (hunk 2 above)
Comment (dlabaj) [reply frontend/src/utilities/time.ts:21]: Updated code to comply with PF guidelines and fixed this issue https://www.patternfly.org/v4/ux-writing/numerics#absolute-or-relative-time

File: frontend/src/utilities/time.ts (hunk 3 above)
Comment (dlabaj) [reply frontend/src/utilities/time.ts:17]: https://www.patternfly.org/v4/ux-writing/numerics#absolute-or-relative-time

File: .env (hunk 1 above)
Comment (dlabaj) [reply .env:18]: removed this

File: backend/src/routes/api/notebook/notebooksUtils.ts [hunk 7]
```diff
@@ -73,31 +96,37 @@ export const getNotebook = async (
...
+        'v1beta1',
+        namespace,
+        'pipelineruns',
+        params.notebook,
+      )
+      .then((r) => r.body as PipelineRunKind)
+      .catch((r) => null);
 
     if (pipelineRun) {
       return { notebooks: mapPipelineRunToNotebook(pipelineRun), error: null };
```
[the remark this answers is not in the document]
Comment (dlabaj) [reply backend/src/routes/api/notebook/notebooksUtils.ts:134]: i think this name is fine here.

**Pull request merged**